### ISSUE TITLE
Automatically closing abandoned issues

### ISSUE THREAD

**Issue by bstriner** [fchollet/keras#5164]
@fchollet I came across something to make life better. There are a handful of tools out there to automatically close issues if they have been abandoned. Here is one example but there are plenty of options.

https://github.com/twbs/no-carrier

A configuration like this might be reasonable:
* Send a warning at 2, 4 and 6 weeks.
* Close after no activity for 8 weeks (other than the warnings), making sure to mark the resolution as something like "automatic closure" so it doesn't get confused with things that are manually closed. It would be easy to browse the automatic closures if you wanted to.
* If something gets closed by accident, someone could always reopen it
* Make some exceptions for any issues you want to pin and keep active.

There are currently 1,945 issues (not including this one) and rising. Does this sound like something reasonable to implement?

Cheers

**Comment by fchollet** [comment]
Yes, sounds good to me.

**Comment by bstriner** [comment]
I'll try it out and setup a demo in a test repository. If it looks good, we would just need to add a cron job to travis.

**Comment by bstriner** [comment]
@fchollet The available tools weren't exactly what we needed, so I just put together a new module. I haven't done the PyPI release yet but I've tried it out and it looks good. Please take a look at the docs or code and let me know if there are any additional features that would be useful.

https://github.com/bstriner/github-bot-close-inactive-issues

The bot posts warnings every `warning_frequency` days starting at `warning_start` days after no activity. The bot closes issues after `closing` days. It can also mark that issue with a label like "closed_automatically" so it will be easy to search. If you run the bot with `--test` it will log the actions it would take but won't actually do anything.

I also put in an option `first_closing_date` so this can be phased in. It will only post warnings and not close any issues (no matter how old) until that date, so people can have a heads-up.

I did some testing in a separate repo. I turned the delay to 0 so it would immediately close the issues so I could test. This is what those automatically closed issues look like:

https://github.com/bstriner/github-bot-test/issues?q=is%3Aissue+is%3Aclosed

In terms of setup:
* Create a new account for the bot, something like `keras-bot`
* Generate a token for the account so it can connect with a high rate limit
* Confirm an email address for the bot so it doesn't get flagged
* Add the bot as a contributor to Keras so it will have permission to close issues
* Edit the configuration file and run `github-close-inactive-issues.py`

**My only concern is how to secure the bot account password.** If the bot password is stored somewhere on the main travis server, someone could possibly get to it. Maybe we could setup travis to be secure but I would feel better if the bot was running on a different machine. Only need to run it every week or so for maintenance. Would you want to setup a local cron job so you know that password can't get out?

**Comment by bstriner** [comment]
Maybe I should add some special handling to ignore issues created by @fchollet and issues with certain labels like help wanted.

Also, regarding setup, you would have to add a new label to the project that we could use for tracking the automatically closed issues.

Cheers

**Comment by bstriner** [comment]
I just ran `--test` against keras. Here is what the results look like. Tons of issues to potentially close.

There are so many issues, I hit the rate limit when I was nearing the end.

The rate resets every hour, so I will add some logic to wait for the rate limit to reset. That means it might take 2-3 hours to run, at least until the number of issues gets down to a manageable number.

> 2017-01-25 02:27:27,540 - root - INFO - Checking rate limit for user github-bot-bot
> 2017-01-25 02:27:27,861 - root - INFO - Limit: 5000, Remaining: 4915, Reset: 2017-01-25 02:56:54
> 2017-01-25 02:27:27,862 - root - INFO - Starting close_inactive_issues
> 2017-01-25 02:27:27,864 - root - INFO - Repo: fchollet/keras, User: github-bot-bot
> 2017-01-25 02:27:27,864 - root - INFO - warning_start: 14, warning_frequency: 7, closing: 56
> 2017-01-25 02:28:11,351 - root - INFO - Issue 4987: warning posted, inactive for 14 days, will be closed after 2017-03-08 03:59:58
> 2017-01-25 02:28:12,773 - root - INFO - Issue 4984: warning posted, inactive for 14 days, will be closed after 2017-03-08 01:16:20
> 2017-01-25 02:28:13,115 - root - INFO - Issue 4982: warning posted, inactive for 14 days, will be closed after 2017-03-07 21:48:54
> 2017-01-25 02:28:13,513 - root - INFO - Issue 4981: warning posted, inactive for 14 days, will be closed after 2017-03-08 00:44:50
> 2017-01-25 02:28:14,171 - root - INFO - Issue 4978: warning posted, inactive for 14 days, will be closed after 2017-03-07 13:50:14
> 2017-01-25 02:28:14,513 - root - INFO - Issue 4977: warning posted, inactive for 14 days, will be closed after 2017-03-07 11:04:31
> 2017-01-25 02:28:14,845 - root - INFO - Issue 4976: warning posted, inactive for 14 days, will be closed after 2017-03-07 10:42:54
> 2017-01-25 02:28:15,733 - root - INFO - Issue 4974: warning posted, inactive for 15 days, will be closed after 2017-03-07 03:59:06
> 2017-01-25 02:28:16,479 - root - INFO - Issue 4969: warning posted, inactive for 15 days, will be closed after 2017-03-06 08:08:11
>

**Comment by bstriner** [comment]
Added some logic to periodically check the remaining rate limit and sleep until it is reset. Should be close to ready now.

**Comment by bstriner** [comment]
With current settings, the script took about 2 hours to run and wants to close 1,809  issues that are inactive. This list will give you an idea of about how long each issue has been inactive.

https://gist.github.com/bstriner/967b45c6d854d257055bcf9d86f44d99

Cheers!

**Comment by bstriner** [comment]
Just released v0.0.2 to PyPI of `github-bot-close-inactive-issues`. I added the ability to ignore issues based on the creator and based on the label. I also fixed the entry points so I can use it on linux and windows.

https://pypi.python.org/pypi?:action=display&name=github-bot-close-inactive-issues
https://github.com/bstriner/github-bot-close-inactive-issues

With this update, you can just `pip install github-bot-close-inactive-issues` and that will put `github-rate-limit` (which just checks the rate limit) and `github-close-inactive-issues` (which closes the issues) onto your path.

So, I would envision the bot skipping anything created by @fchollet and anything with help wanted or bug as the label.

Even if we only automatically close issues after six months, that will still knock a ton of issues out of Keras.

Are there any other features or contingencies to consider?

Cheers

**Comment by patyork** [comment]
Kind of off topic, but I've had luck parring down the people that post multiple implementation questions as issues using the below code if/when I comment:

```
<br />
---

If your issue is an implementation question, please ask your question on [StackOverflow](http://stackoverflow.com/questions/tagged/keras) or [join the Keras Slack channel](https://keras-slack-autojoin.herokuapp.com/) and ask there instead of filing a GitHub issue.
```

Looks like the below. People generally aren't reading the stuff that is in the issue template, but (unsurprisingly) they *do* read answer people give them ;)

<br />

---

If your issue is an implementation question, please ask your question on [StackOverflow](http://stackoverflow.com/questions/tagged/keras) or [join the Keras Slack channel](https://keras-slack-autojoin.herokuapp.com/) and ask there instead of filing a GitHub issue.

**Comment by bstriner** [comment]
@patyork not off topic at all. That text or something similar should be in the warning for the bot.

https://github.com/bstriner/github-bot-close-inactive-issues/blob/master/github_bot.yml

I've started playing with the Github API and you could automate whatever you want. Crazy idea: use Github labels to mark issues as bugs vs. implementation questions and train a classifier to do some automation.

Google group also seems like a good place to talk. I'm not a huge fan of slack yet. Maybe include SO, slack and google group in the message?

**Comment by bstriner** [comment]
@fchollet I did some poking around on the best way to configure something like this. Create a separate github repo with just a travis.yml and a github bot configuration file. Enable travis, disable builds on pull requests and enable cron builds. Configure the bot through travis environment variables.

The travis environment variables will be hidden to other users and they will not be able to write malicious code to access the variables if you disable builds on pull requests. That means the bot needs to run on a separate travis instance than the rest of keras.

Seems like it should be reasonable to get the ball rolling on the backlog. Still interested?

**Comment by fchollet** [comment]
Yes, I still think this would be a good feature to have. Albeit it should
be paired with a reasonable level of formal handling of Github issues
(currently relatively few people look at them, and not in a structured
fashion).

**Comment by bstriner** [comment]
@fchollet You could configure the bot to just post warnings on stale issues and not actually close anything. That way you have some automated way to keep the ball rolling on issues and see if they are still active without accidentally closing something legit.

The bot could be a good start but there has to be a better way to handle issues. There are some actual legit issues buried in stacks of questions and there is no way to get a good handle on what is what.

It would be nice to use tags to differentiate, but you either do that all yourself or give other people the ability to nuke keras.

Github permissions are junk. It is basically all or none. I'm thinking about making some other apps to help with this kind of issue (not keras-specific).

Automatic closing and better permissions should be a part of Github. It would be like one line of SQL. However, while those things aren't in Github, it is something that can be built using the API.

What would you think of a Github permissions app? You would be able to add users and give them fine-grained permissions like adding tags, removing tags, closing issues, etc. The app would store the permissions table, and proxy commands to an owner account if they pass the permissions check.

The users view issues and PRs through the proxy app. Commands to comment go through their account. Commands to close, tag, merge, or whatever are proxied to an owner account that posts a comment like "closed on behalf of [username]" if permissions pass.

The bot for warnings and closings on issues could be rolled into the app as well.

Does that seem like something that would actually be useful on a broader scale? Wouldn't be hard to put something together if I can find some cheap hosting.

Cheers

**Comment by bstriner** [comment]
@fchollet Glad you got something up and running. probot-stale looks like a javascript clone of my python repo but at least I don't have to support anything this way. It looks like they have some nice features. Please let me know any thoughts on how it works out.

Cheers
